packages/app/src/cli/services/function/build.ts

```diff
@@ -137,3 +157,121 @@ export async function runFunctionRunner(fun: ExtensionInstance, options: Functio
     stderr: 'inherit',
   })
 }
+
+export interface JavyBuilder {
+  bundle(
+    fun: ExtensionInstance<FunctionConfigType>,
+    options: JSFunctionBuildOptions,
+  ): Promise<BuildResult<BuildOptions>>
+  compile(fun: ExtensionInstance<FunctionConfigType>, options: JSFunctionBuildOptions): Promise<void>
+}
+
+export const DefaultJavyBuilder: JavyBuilder = {
+  async bundle(fun: ExtensionInstance<FunctionConfigType>, options: JSFunctionBuildOptions) {
+    return bundleExtension(fun, options)
+  },
+
+  async compile(fun: ExtensionInstance<FunctionConfigType>, options: JSFunctionBuildOptions) {
+    return runJavy(fun, options)
+  },
+}
+
+const JAVY_WORLD = 'shopify-function'
+export class ExportJavyBuilder implements JavyBuilder {
+  exports: string[]
+
+  constructor(exports: string[]) {
+    this.exports = exports
+  }
+
+  async bundle(fun: ExtensionInstance<FunctionConfigType>, _options: JSFunctionBuildOptions) {
+    const javyFs = await findPathUp('node_modules/javy/fs', {
+      type: 'directory',
+      cwd: fun.directory,
+    })
+    if (!javyFs) {
+      throw new Error("Could not find the javy/fs runtime. Make sure you have 'javy/fs' installed")
+    }
+
+    if (!fun.entrySourceFilePath) {
+      throw new Error('Could not find your function entry point. It must be in src/index.js or src/index.ts')
+    }
+
+    const contents = this.entrypointContents
+    outputDebug('Generating dist/function.js using generated module:')
+    outputDebug(contents)
+
+    const esbuildOptions: Parameters<typeof esBuild>[0] = {
+      outfile: joinPath(fun.directory, 'dist/function.js'),
+      stdin: {
+        contents,
+        loader: 'ts',
+      },
+      alias: {
+        'user-function': fun.entrySourceFilePath,
+        'javy/fs': javyFs,
+      },
+      logLevel: 'silent',
+      bundle: true,
+      legalComments: 'none',
+      target: 'es2022',
+      format: 'esm',
+    }
+
+    return esBuild(esbuildOptions)
+  }
+
+  async compile(fun: ExtensionInstance<FunctionConfigType>, options: JSFunctionBuildOptions) {
+    const witPath = joinPath(fun.directory, 'dist', 'javy-world.wit')
+    await writeFile(witPath, this.wit)
+    const javyArgs = ['--wit', witPath, '-n', JAVY_WORLD]
+
+    return runJavy(fun, options, javyArgs)
+  }
+
+  get wit() {
+    // % escapes the name to avoid conflict with reserved words, if any
+    const witExports = this.exports.map((name) => `export %${hyphenate(name)}: func()`)
+    return `package function:impl
+
+world ${JAVY_WORLD} {
+  ${witExports.join('\n  ')}
+}`
+  }
+
+  get entrypointContents() {
+    const prelude = `
+import * as userExports from "user-function"
+import * as fs from "javy/fs"
+
+function run(fn) {
+  const input_data = fs.readFileSync(fs.STDIO.Stdin);
+  const input_str = new TextDecoder("utf-8").decode(input_data);
+  const input_obj = JSON.parse(input_str);
+  const output_obj = userfunction(input_obj);
```

Comment:
What is this `userfunction`? `userFunction`?

Reply:
It should read `fn`, sorry. I'll update soon with a change in `@shopify/shopify_function`.